### PR TITLE
Add basic support for MRV OptiDriver Optical Transport Platform

### PR DESCRIPTION
**Pull request by thecityofguanyu**
Add basic support for MRV OptiDriver Optical Transport Platform
 * OS Detection
 * Sensors
  * Chassis temperature
  * State
   * Tx/Rx Optical Interface Power
   * Power Supply
   * Fan

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [ ] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

### CODE REVIEW

File: includes/definitions/mrv-optidriver.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+os: mrv-optidriver
```
Comment (laf) [inline includes/definitions/mrv-optidriver.yaml]: Is this the best os name?

File: includes/definitions/mrv-optidriver.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+group: mrv
```
Comment (laf) [inline includes/definitions/mrv-optidriver.yaml]: Do you need this? It's designed to bring a number of OS' together so polling can be done centrally.

File: includes/definitions/mrv-optidriver.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+text: 'MRV OptiDriver'
+type: network
+icon: mrv
+discovery:
+    - sysObjectId:
+       - .1.3.6.1.4.1.629.200.1.1.1.
```
Comment (laf) [inline includes/definitions/mrv-optidriver.yaml]: You might be able to change this to `.1.3.6.1.4.1.629.200.`

File: includes/discovery/sensors/dbm/mrv-optidriver.inc.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
+
+echo 'MRV OptiDriver ';
+
+$multiplier = 1;
+$divisor = 1000;
+$limit_low = -100;
```
Comment (laf) [inline includes/discovery/sensors/dbm/mrv-optidriver.inc.php]: Are these limits sane?

File: includes/discovery/sensors/dbm/mrv-optidriver.inc.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+foreach ($pre_cache['mrv-optidriver_oids'] as $index => $entry) {
+    $oid = '.1.3.6.1.4.1.629.200.8.1.1.32.' . $index;
+    $descr = dbFetchCell('SELECT `ifDescr` FROM `ports` WHERE `ifName`= ? AND `device_id` = ?', array($index, $device['device_id'])) . ' Rx Power';
+    $currentrx = $entry['nbsCmmcPortRxPower'];
+    discover_sensor(
+        $valid['sensor'],
+        'dbm',
+        $device,
+        $oid,
+        'rx-' . $index,
```
Comment (laf) [inline includes/discovery/sensors/dbm/mrv-optidriver.inc.php]: probably better to do `'nbsCmmcPortRxPower.'.$index,`

File: includes/discovery/sensors/dbm/mrv-optidriver.inc.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        $limit_low,
+        $warn_limit_low,
+        $warn_limit,
+        $limit,
+        $currentrx,
+        'snmp'
+    );
+}
+
+foreach ($pre_cache['mrv-optidriver_oids'] as $index => $entry) {
```
Comment (laf) [inline includes/discovery/sensors/dbm/mrv-optidriver.inc.php]: You're already looping through this array above, you can probably do an if / elseif check to decide which value you want and consolidate the code down a bit.

File: includes/discovery/sensors/dbm/mrv-optidriver.inc.php [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    $oid = '.1.3.6.1.4.1.629.200.8.1.1.31.' . $index;
+    $descr = dbFetchCell('SELECT `ifDescr` FROM `ports` WHERE `ifName`= ? AND `device_id` = ?', array($index, $device['device_id'])) . ' Tx Power';
+    $currenttx = $entry['nbsCmmcPortTxPower'];
+    discover_sensor(
+        $valid['sensor'],
+        'dbm',
+        $device,
+        $oid,
+        'tx-' . $index,
```
Comment (laf) [inline includes/discovery/sensors/dbm/mrv-optidriver.inc.php]: probably better to do 'nbsCmmcPortTxPower.'.$index,

File: includes/discovery/sensors/state/mrv-optidriver.inc.php [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
+
+<?php
+
+// Optical Interface Power State
+if ($device['os'] == 'mrv-optidriver') {
```
Comment (laf) [inline includes/discovery/sensors/state/mrv-optidriver.inc.php]: You don't need this check anymore.

File: includes/discovery/sensors/state/mrv-optidriver.inc.php [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+    $oids = snmpwalk_cache_multi_oid($device, 'nbsCmmcPortTable', array(), 'NBS-CMMC-MIB');
```
Comment (laf) [inline includes/discovery/sensors/state/mrv-optidriver.inc.php]: You've looked this up in pre-cache so re-use it.

File: includes/discovery/sensors/state/mrv-optidriver.inc.php [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+                    'state_descr' => $value[1],
+                    'state_draw_graph' => $value[2],
+                    'state_value' => $value[3],
+                    'state_generic_value' => $value[4]
+                );
+                dbInsert($insert, 'state_translations');
+            }
+        }
+
+        foreach ($oids as $index => $entry) {
```
Comment (laf) [inline includes/discovery/sensors/state/mrv-optidriver.inc.php]: Again you can do most of this within the first loop of the data.

File: includes/discovery/sensors/state/mrv-optidriver.inc.php [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+
+            //Create Sensor To State Index
+            create_sensor_to_state_index($device, $state_name, $index);
+        }
+    }
+}
+
+
+// Power Supply State
+if ($device['os'] == 'mrv-optidriver') {
```
Comment (laf) [inline includes/discovery/sensors/state/mrv-optidriver.inc.php]: This check isn't needed

File: includes/discovery/sensors/state/mrv-optidriver.inc.php [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+        }
+
+        //Discover Sensors
+        discover_sensor($valid['sensor'], 'state', $device, $cur_oid, $index, $state_name, 'Power Supply 1', '1', '1', null, null, null, null, $oids['nbsCmmcChassisPS1Status'], 'snmp', 1);
+
+        //Create Sensor To State Index
+        create_sensor_to_state_index($device, $state_name, $index);
+    }
+    // PS2
+    if (is_array($oids)) {
```
Comment (laf) [inline includes/discovery/sensors/state/mrv-optidriver.inc.php]: Again you can do most of this within the first loop of the data.

File: includes/discovery/sensors/state/mrv-optidriver.inc.php [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+        }
+
+        //Discover Sensor
+        discover_sensor($valid['sensor'], 'state', $device, $cur_oid, $index, $state_name, 'Power Supply 2', '1', '1', null, null, null, null, $oids['nbsCmmcChassisPS2Status'], 'snmp', 1);
+
+        //Create Sensor To State Index
+        create_sensor_to_state_index($device, $state_name, $index);
+    }
+    // PS3
+    if (is_array($oids)) {
```
Comment (laf) [inline includes/discovery/sensors/state/mrv-optidriver.inc.php]: Again you can do most of this within the first loop of the data.

File: includes/discovery/sensors/state/mrv-optidriver.inc.php [hunk 14]
```diff
@@ -0,0 +1,408 @@
...
+        }
+
+        //Discover Sensor
+        discover_sensor($valid['sensor'], 'state', $device, $cur_oid, $index, $state_name, 'Power Supply 3', '1', '1', null, null, null, null, $oids['nbsCmmcChassisPS3Status'], 'snmp', 1);
+
+        //Create Sensor To State Index
+        create_sensor_to_state_index($device, $state_name, $index);
+    }
+    // PS4
+    if (is_array($oids)) {
```
Comment (laf) [inline includes/discovery/sensors/state/mrv-optidriver.inc.php]: Again you can do most of this within the first loop of the data.

File: includes/discovery/sensors/state/mrv-optidriver.inc.php [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+
+        //Discover Sensor
+        discover_sensor($valid['sensor'], 'state', $device, $cur_oid, $index, $state_name, 'Power Supply 4', '1', '1', null, null, null, null, $oids['nbsCmmcChassisPS4Status'], 'snmp', 1);
+
+        //Create Sensor To State Index
+        create_sensor_to_state_index($device, $state_name, $index);
+    }
+
+    // Fan 1
+    if (is_array($oids)) {
```
Comment (laf) [inline includes/discovery/sensors/state/mrv-optidriver.inc.php]: Same as the psu ones above.

File: includes/discovery/sensors/temperature/mrv-optidriver.inc.php [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
+
+
+if ($device['os'] == 'mrv-optidriver') {
```
Comment (laf) [inline includes/discovery/sensors/temperature/mrv-optidriver.inc.php]: This check isn't needed

File: includes/discovery/sensors/temperature/mrv-optidriver.inc.php [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    echo("MRV OptiDriver:");
+    // Chassis temperature
+    $high_limit         = 70;
```
Comment (laf) [inline includes/discovery/sensors/temperature/mrv-optidriver.inc.php]: Are these sane limits?

File: includes/discovery/sensors/temperature/mrv-optidriver.inc.php [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    $high_warn_limit    = 65;
+    $low_warn_limit     = 5;
+    $low_limit          = 0;
+    $descr   =  "Chassis Temperature";
+    $valueoid    = ".1.3.6.1.4.1.629.200.6.1.1.15.1";
+    $value   = snmp_get($device, 'nbsCmmcChassisTemperature.1', '-Ovqs', 'NBS-CMMC-MIB', $config['install_dir'].'/mibs/mrv');
+    //$value = str_replace('"', '', $value);
+    if (is_numeric($value)) {
+        discover_sensor($valid['sensor'], 'temperature', $device, $valueoid, 1, 'mrv-optidriver', $descr, '1', '1', $low_limit, $low_warn_limit, $high_warn_limit, $high_limit, $value);
```
Comment (laf) [inline includes/discovery/sensors/temperature/mrv-optidriver.inc.php]: Change the index of 1 to be `nbsCmmcChassisTemperature.1`

File: includes/polling/os/mrv-optidriver.inc.php [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+ * © 2017 Chris A. Evans <thecityofguanyu@outlook.com>
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
+
+$hardware = snmp_get($device, 'nbsCmmcChassisModel.1', '-Ovqs', 'NBS-CMMC-MIB', $config['install_dir'].'/mibs/mrv');
```
Comment (laf) [inline includes/polling/os/mrv-optidriver.inc.php]: Please convert all of these to use a single snmp_mult_get_oid call

File: includes/definitions/mrv-optidriver.yaml (hunk 2 above)
Comment (laf) [inline includes/definitions/mrv-optidriver.yaml]: If it is removed, add:

```
mib_dir:
    - mrv
```

**Comment by laf** [comment]
You've got numerous issues that need fixing in the code to make it PSR-2 compliant: https://travis-ci.org/librenms/librenms/jobs/232922495

Also needs unit tests.

File: includes/definitions/mrv-optidriver.yaml (hunk 2 above)
Comment (thecityofguanyu) [inline includes/definitions/mrv-optidriver.yaml]: @laf I based that off of the pre-existing `./includes/definitions/mrvld.yaml`, which denotes the `mrv` group. MRV has a larger suite of devices than just OptiDriver (OptiPacket, OptiSwitch), so I would guess that they could have definition OS types.

Comment (laf) [inline includes/definitions/mrv-optidriver.yaml]: It's only needed if they share the same code which they don't appear to.

File: includes/definitions/mrv-optidriver.yaml (hunk 3 above)
Comment (thecityofguanyu) [inline includes/definitions/mrv-optidriver.yaml]: The 1.1.1 is what's showing in the full snmpget of sysObjectID.

```
.1.3.6.1.2.1.1.2.0 = OID: .1.3.6.1.4.1.629.200.1.1.1
SNMPv2-MIB::sysObjectID.0 = OID: NBS-CMMC-MIB::nbsCmmcObjects.1.1
```

Still want it shorted?

File: includes/discovery/sensors/dbm/mrv-optidriver.inc.php (hunk 4 above)
Comment (thecityofguanyu) [inline includes/discovery/sensors/dbm/mrv-optidriver.inc.php]: Multiplier and divisor are right, but those limits are definitely not sane. I'll see if I can find some documentation to set those accordingly.

File: includes/definitions/mrv-optidriver.yaml (hunk 3 above)
Comment (laf) [inline includes/definitions/mrv-optidriver.yaml]: Looking at our stats, nothing else uses it so yes it should be ok to be shortened.

File: includes/discovery/sensors/dbm/mrv-optidriver.inc.php (hunk 4 above)
Comment (laf) [inline includes/discovery/sensors/dbm/mrv-optidriver.inc.php]: You can set them to null and we will calculate some values.

File: includes/discovery/sensors/state/mrv-optidriver.inc.php (hunk 9 above)
Comment (thecityofguanyu) [inline includes/discovery/sensors/state/mrv-optidriver.inc.php]: Could I do `$oids = $pre_cache['mrv-od_oids']` here in lieu of the existing statement?

File: includes/discovery/sensors/dbm/mrv-optidriver.inc.php (hunk 6 above)
Comment (thecityofguanyu) [inline includes/discovery/sensors/dbm/mrv-optidriver.inc.php]: Is there a similar existing file using an if / elif in this manner that I could reference? I'm struggle to figure out what the conditional would be.

Comment (thecityofguanyu) [inline includes/discovery/sensors/dbm/mrv-optidriver.inc.php]: `if (is_numeric($entry['nbsCmmcPortRxPower']) {` and `if (is_numeric($entry['nbsCmmcPortTxPower']) {` is what I have come up with thus far.

**Comment by thecityofguanyu** [comment]
@laf I believe that I have addressed for of your optimization / tidy up issues in the latest commit. I have not yet addressed the Travis complaints. I want to ensure the code is all how you want it before I touch on that.

File: includes/definitions/mrv-od.yaml [hunk 20]
```diff
@@ -0,0 +1,11 @@
+os: mrv-od
+text: 'MRV OptiDriver'
+type: network
+icon: mrv
+mib_dir:
+    - mrv
+discovery:
+    - sysObjectId:
+       - .1.3.6.1.4.1.629.200.
+    - sysDescr:
```
Comment (laf) [inline includes/definitions/mrv-od.yaml]: You can probably drop the sysDescr check

File: includes/discovery/sensors/temperature/mrv-od.inc.php [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
+
+
+echo("MRV OptiDriver:");
+// Chassis temperature
+$descr   =  "Chassis Temperature";
+$valueoid    = ".1.3.6.1.4.1.629.200.6.1.1.15.1";
+$value   = snmp_get($device, 'nbsCmmcChassisTemperature.1', '-Ovqs', 'NBS-CMMC-MIB', $config['install_dir'].'/mibs/mrv');
```
Comment (laf) [inline includes/discovery/sensors/temperature/mrv-od.inc.php]: You shouldn't need (or use) ``$config['install_dir'].'/mibs/mrv'`. Our code will populate that.

File: includes/polling/os/mrv-od.inc.php [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
+
+$multi_get_array = snmp_get_multi($device, 'nbsCmmcChassisModel.1 nbsCmmcSysFwVers.0 nbsCmmcChassisSerialNum.1', '-Ovqs', 'NBS-CMMC-MIB', $config['install_dir'].'/mibs/mrv');
+$hardware = $multi_get_array[1]['NBS-CMMC-MIB::nbsCmmcChassisModel'];
+$version = $multi_get_array[0]['NBS-CMMC-MIB::nbsCmmcSysFwVers'];
+$serial = $multi_get_array[1]['NBS-CMMC-MIB::nbsCmmcChassisSerialNum'];
+$features       = '';
```
Comment (laf) [inline includes/polling/os/mrv-od.inc.php]: Add `unset($multi_get_array);`

**Pull request merged**